aws/aws-parallelcluster
# Improve Existing EFS, FSx Security Groups Validation

In addition to existing checks on port and CIDR, this commit adds checks on allowed security groups. This improvement will fix false warning during `pcluster update-cluster`. 

In other words, there are three ways to specify src/dst in a security group rule: CIDR, security groups, prefix list. Before this commit, only CIDR was properly checked. This commit adds proper checks for security groups. This commit does not change the behavior of prefix list .

This commit also refactors the code for clarity and reduces redundant EFS error messages

Signed-off-by: Hanwen <hanwenli@amazon.com>

### Tests
* Unit tests have been created and modified
#### Manual testing:
When creating clusters with bad EFS and FSx security groups, the error messages are correctly displayed and only appear once:
```
{
  "configurationValidationErrors": [
    {
      "level": "ERROR",
      "type": "EfsIdValidator",
      "message": "There is an existing Mount Target fsmt-xxx in the Availability Zone us-east-1b for EFS fs-xxx, but it does not have a security group that allows inbound and outbound rules to support NFS. Please modify the Mount Target's security group, to allow traffic on port 2049."
    },
    {
      "level": "ERROR",
      "type": "ExistingFsxNetworkingValidator",
      "message": "The current security group settings on file system 'fs-xxx' does not satisfy mounting requirement. The file system must be associated to a security group that allows inbound and outbound TCP traffic through ports [988]. Missing ports: [988]"
    }
  ],
  "message": "Invalid cluster configuration."
}
```
When updating cluster with EFS and FSx with security groups allowing connections from cluster security groups, the validator no longer shows false error messages
```
{
  "message": "Request would have succeeded, but DryRun flag is set.",
  "changeSet": [
    {
      "parameter": "SharedStorage",
      "requestedValue": {
        "MountDir": "/efs1",
        "Name": "efs1",
        "StorageType": "Efs",
        "EfsSettings": {
          "FileSystemId": "fs-01cf0078e64e83b54"
        }
      },
      "currentValue": "-"
    },
    {
      "parameter": "SharedStorage",
      "requestedValue": {
        "MountDir": "/fsx1",
        "Name": "fsx1",
        "StorageType": "FsxLustre",
        "FsxLustreSettings": {
          "FileSystemId": "fs-01fef606ec594d3b3"
        }
      },
      "currentValue": "-"
    }
  ]
}

```

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
